lib/golem/state_estimator.h

```diff
@@ -0,0 +1,79 @@
+#ifndef GOLEM_STATE_ESTIMATOR_H
+#define GOLEM_STATE_ESTIMATOR_H
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace golem {
+
+// Models the state estimator of your system
+// Accepts any class that implements the following functions for given State and Feedback types:
+// - void update(const Feedback&) const;
+// - State get() const;
+template <typename State, typename Feedback>
+class StateEstimator {
+public:
+    StateEstimator() = default;
+    StateEstimator(const StateEstimator&) = delete;
+    StateEstimator& operator=(const StateEstimator&) = delete;
+    StateEstimator(StateEstimator&&) = default;
+    StateEstimator& operator=(StateEstimator&&) = default;
+
+    template <typename T>
+    StateEstimator(T&& impl)
+        : m_impl(new model_t<typename std::decay<T>::type>(std::forward<T>(impl)))
+    {
+    }
+
+    template <typename T>
+    StateEstimator& operator=(T&& impl)
+    {
+        m_impl.reset(new model_t<typename std::decay<T>::type>(std::forward<T>(impl)));
```

Comment:
Same as before, not sure if I like malloc here.

Reply:
bye bye malloc, hello static polymorphism